runtime/common/src/evm/fees.rs

```diff
@@ -0,0 +1,130 @@
+
+impl<R> OnChargeEVMTransaction<R> for DealWithEVMFees<R>
+where
+	R: pallet_balances::Config
+		+ pallet_treasury::Config
+		+ pallet_authorship::Config
+		+ pallet_evm::Config,
+	Treasury<R>: OnUnbalanced<NegativeImbalance<R>>,
+	Balances<R>: Currency<<R as frame_system::Config>::AccountId>,
+	U256: UniqueSaturatedInto<Balance<R>>,
+{
+	type LiquidityInfo = Option<NegativeImbalance<R>>;
+
+	fn withdraw_fee(who: &H160, fee: U256) -> Result<Self::LiquidityInfo, Error<R>> {
+		if fee.is_zero() {
+			return Ok(None);
+		}
+		let account_id = R::AddressMapping::into_account_id(*who);
+		Balances::<R>::withdraw(
+			&account_id,
+			fee.unique_saturated_into(),
+			WithdrawReasons::FEE,
+			ExistenceRequirement::AllowDeath,
+		)
+		.map(Some)
+		.map_err(|_| Error::<R>::BalanceLow)
+	}
+
+	fn correct_and_deposit_fee(
+		who: &H160,
+		corrected_fee: U256,
+		base_fee: U256,
+		already_withdrawn: Self::LiquidityInfo,
+	) -> Self::LiquidityInfo {
+		let Some(paid) = already_withdrawn else { return None };
+
+		let account_id = R::AddressMapping::into_account_id(*who);
+
+		// Calculate how much refund we should return
+		let refund_amount = paid
+			.peek()
+			.saturating_sub(corrected_fee.unique_saturated_into());
+		// refund to the account that paid the fees. If this fails, the
+		// account might have dropped below the existential balance. In
+		// that case we don't refund anything.
+		let refund_imbalance = Balances::<R>::deposit_into_existing(&account_id, refund_amount)
+			.unwrap_or_else(|_| PositiveImbalance::<R>::zero());
+
+		// Make sure this works with 0 ExistentialDeposit
+		// https://github.com/paritytech/substrate/issues/10117
+		// If we tried to refund something, the account still empty and the ED is set to
+		// 0, we call `make_free_balance_be` with the refunded amount.
+		let refund_imbalance = if Balances::<R>::minimum_balance().is_zero()
+			&& refund_amount > <Balance<R>>::zero()
+			&& Balances::<R>::total_balance(&account_id).is_zero()
+		{
+			// Known bug: Substrate tried to refund to a zeroed AccountData, but
+			// interpreted the account to not exist.
+			match Balances::<R>::make_free_balance_be(&account_id, refund_amount) {
+				SignedImbalance::Positive(p) => p,
+				_ => PositiveImbalance::<R>::zero(),
+			}
+		} else {
+			refund_imbalance
+		};
+
+		// merge the imbalance caused by paying the fees and refunding parts of it
+		// again.
+		let adjusted_paid = paid
+			.offset(refund_imbalance)
+			.same()
+			.unwrap_or_else(|_| NegativeImbalance::<R>::zero());
+
+		let (base_fee, tip) = adjusted_paid.split(base_fee.unique_saturated_into());
+
+		// Handle base fee and tips
+		let (treasury_amount, mut author_amount) = base_fee.ration(
+			TREASURY_FEE_RATIO.deconstruct(),
+			(Perbill::one() - TREASURY_FEE_RATIO).deconstruct(),
+		);
+		tip.merge_into(&mut author_amount);
+		<Treasury<R> as OnUnbalanced<_>>::on_unbalanced(treasury_amount);
+		<ToAuthor<R> as OnUnbalanced<_>>::on_unbalanced(author_amount);
```

Comment:
This section reminds me to `DealWithFees` type, located in `runtime/common/src/lib.rs`. I think the structure is the same. Maybe we can call it to avoid duplicate code:

```rust
DealWithFees::on_unbalanceds([base_fee, tip].iter());
```

Or similar.

Reply:
:+1: yeah, originally I had the tip handling done in `pay_priority_fee`, but since I rolled it all into one place it makes sense to call up into `DealWithFees` so we only have one place to update